Improve package schema validation

chammp opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Properly handle validation errors occuring for packages inside core yaml instead of failing at Home Assistant startup. Currently, on package schema validation errors, HA startup completely fails. After this PR, HA will instead load each package that is defined correctly and skip over packages that have a definition error. (This change does not affect definitions _inside_ packages, but the basic package definition instead.) If the whole `packages` node is invalid, no packages will be loaded at all, but HA still starts.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

chammp committed:
Add failing tests for package config validation error wrapping

chammp committed:
Wrap package schema validation errors in HomeAssistantError

MartinHjelmare commented on homeassistant/config.py:
```diff
@@ -1030,7 +1030,11 @@ async def merge_packages_config(
+    try:
+        PACKAGES_CONFIG_SCHEMA(packages)
+    except vol.Error as err:
+        raise HomeAssistantError(f"Error validating package config: {err}") from err
```
I think this function is expected to use `_log_pkg_error` to record package config errors. See example on line 1045.

Since this validation validates all of the packages it may be hard to determine which package(s) is invalid. If we can't do that, I think the caller of `merge_packages_config` should handle `vol.Invalid` and record an error for the whole packages item as the caller does for other config items.

chammp replied:
OK... My rationale was to keep the current behavior (which would simply bail out of the whole config loading) and make it compatible to the currently existing activation of recovery mode, which is supposedly intended as that is what happens for other severe errors on config parse/load. The bootstrap code fetches and treats `HomeAssistantError` and activates recovery mode; on other exceptions it just fails and aborts (see https://github.com/home-assistant/core/blob/6bc36666b1551c9fdc567b02d122398fd687e787/homeassistant/bootstrap.py#L148).

I did explicitly _not_ want to alter the startup behavior. Clearly there are alternatives; for instance we could simply skip the whole "packages" section with a nice friendly error message in the logs (and not activate recovery mode at all), or even follow @emontnemery's suggestion and remove the global schema check of the full packages config and move it inside the for loops to only skip erroneous packages and continue with the others. I am open for both alternatives, I'd just like to know what would be more acceptable, and which checks there are in place besides key_needs_to_be_a_slug and value_needs_to_be_a_dict (if any). I was unable to cause other validation errors, but you can't be too sure ;)

MartinHjelmare replied:
Bootstrap isn't the only place where `merge_packages_config` is indirectly used. It's also used in the check config helper. We need to keep the intended behavior of all users of `merge_packages_config`.

MartinHjelmare replied:
I think the `PACKAGES_CONFIG_SCHEMA` is there to make subsequent validation easier by validating that the main structure of the packages config has the correct types and format.

It will be hard to validate the parts if some baseline isn't established, eg that the container is a dict. If the config container is completely incorrect we need to abort early anyway.

chammp replied:
> Bootstrap isn't the only place where `merge_packages_config` is indirectly used. It's also used in the check config helper. We need to keep the intended behavior of all users of `merge_packages_config`.

I know. Without patch:
![grafik](https://github.com/home-assistant/core/assets/57918757/f7290118-7ddf-48aa-bb3f-728268acbb06)

With patch:
![grafik](https://github.com/home-assistant/core/assets/57918757/f29423a7-07ad-40a2-be93-19d1fd069a1d)

Works for me 😉 

Startup log without patch (I skipped traceback elements to keep it short; you get the idea):
```
  File "/workspaces/homeassistant-core/homeassistant/runner.py", line 154, in setup_and_run_hass
    hass = await bootstrap.async_setup_hass(runtime_config)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspaces/homeassistant-core/homeassistant/bootstrap.py", line 147, in async_setup_hass
    config_dict = await conf_util.async_hass_config_yaml(hass)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspaces/homeassistant-core/homeassistant/config.py", line 505, in async_hass_config_yaml
    await merge_packages_config(hass, config, core_config.get(CONF_PACKAGES, {}))
  File "/workspaces/homeassistant-core/homeassistant/config.py", line 1033, in merge_packages_config
    PACKAGES_CONFIG_SCHEMA(packages)
  File "/workspaces/homeassistant-core/homeassistant/helpers/config_validation.py", line 588, in verify
    return cast(dict, schema(value))
                      ^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/voluptuous/schema_builder.py", line 272, in __call__
    return self._compiled([], data)
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/voluptuous/schema_builder.py", line 595, in validate_dict
    return base_validate(path, iteritems(data), out)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/voluptuous/schema_builder.py", line 433, in validate_mapping
    raise er.MultipleInvalid(errors)
voluptuous.error.MultipleInvalid: expected a dictionary for dictionary value @ data['horst']
```


Startup log with patch:
```
2024-01-17 06:01:03.404 ERROR (MainThread) [homeassistant.bootstrap] Failed to parse configuration.yaml: Error validating package config: expected a dictionary for dictionary value @ data['horst']. Activating recovery mode
```


> It will be hard to validate the parts if some baseline isn't established, eg that the container is a dict. If the config container is completely incorrect we need to abort early anyway.

That depends on what is actually verified. I know of three properties:
1. `packages` is a `dict`
2. Eeach key of `packages` is a slug
3. Each value of `packages` is a dict

To check these properties kind of more explicitely does not seem too complicated.

emontnemery replied:
> My rationale was to keep the current behavior (which would simply bail out of the whole config loading) and make it compatible to the currently existing activation of recovery mode, which is supposedly intended as that is what happens for other severe errors on config parse/load. 

We only active recovery mode automatically for severe issues with the configuration, for example failure to setup frontend.

A problem with package configuration which does not otherwise block startup does not warrant entering recovery mode.

Hence my request to start with improving the error message if `PACKAGES_CONFIG_SCHEMA` raises, and if that's difficult, modify `PACKAGES_CONFIG_SCHEMA`.

Did you look into the functions I mentioned which help format `vol.Invalid` to make it user readable?

chammp replied:
> We only active recovery mode automatically for severe issues with the configuration, for example failure to setup frontend.
> 
> A problem with package configuration which does not otherwise block startup does not warrant entering recovery mode.

I guess that makes sense; however, currently the whole application doesn't start at all in these cases. So having it run in recovery mode is at least an improvement 😉 


> Hence my request to start with improving the error message if `PACKAGES_CONFIG_SCHEMA` raises, and if that's difficult, modify `PACKAGES_CONFIG_SCHEMA`.

I started messing with the validation a little more. One issue is that the control flow between bootstrap and config check is quite different (something I don't plan to address in this MR, though...): while bootstrapping package merge happens before an additional validation against the full core config schema, during config check it's the other way around. There the full schema validation happens first and afterwards package merge is performed.

That said, my current local WIP version looks like this:
1. Simplified `PACKAGES_CONFIG_SCHEMA` that bails out during package merge only on super severe validation errors (which would then trigger recovery mode; yet I was unable to provoke this; don't exactly know yet what I can do to my config files in order to actually trigger _that_) (that also affects `CORE_CONFIG_SCHEMA` btw, but as this is always used together with merge in some way, I think it's fine)
2. Check each key in `packages` for being a slug and each value being a dict and produce nice sweet error messages in case something goes wrong; something like `2024-01-17 15:17:21.716 ERROR (MainThread) [homeassistant.config] Setup of package 'hos t' at packages/System/Host.yaml, line 1 failed: Invalid package definition 'hos t': invalid slug hos t (try hos_t). Package will not be initialized` during bootstrap and config reload.
3. Ignore packages that fail this validation, but loads the rest

Config reload check produces a message like this:
![grafik](https://github.com/home-assistant/core/assets/57918757/4cec234f-1bad-4707-b972-fafa454457b3)

Actual config reload adds this message to system log.

How do you feel about this?

chammp replied:
I updated the code and the title and description of this PR to basically reflect what I wrote in my last message.

chammp committed:
Fix yamllint errors

chammp committed:
Rework package merge validation

Ignore invalid package definitions instead of failing startup.
Output error messages with locations if possible when a package
definition has errors.

chammp committed:
Ruff format

chammp committed:
Fix linter errors

chammp commented:
Regarding this CI failure
![grafik](https://github.com/home-assistant/core/assets/57918757/1d48bc85-7c70-48c0-88eb-24cc23632926)
I think this is unrelated to my changes. 

![grafik](https://github.com/home-assistant/core/assets/57918757/ecf466bc-727f-45ac-9336-f4a6c6e9f1d0)

The test that fails is not related to this PR; I did my best to forward-trace the code execution. Just FYI, dear reviewer(s) 😉

emontnemery commented on homeassistant/config.py:
```diff
@@ -1023,12 +1031,38 @@ async def merge_packages_config(
+    Raises vol.Invalid if whole package config is invalid.
+    """
+
+    package_definition_schema = vol.Schema({cv.string: vol.Any(dict, list, None)})
```
Let's move this to module level

chammp replied:
Done.

emontnemery reviewed (commented):
Very nice @chammp 👍 

Just a minor remark about moving the local `package_definition_schema` to module level

chammp committed:
Move package_definition_schema to module scope

MartinHjelmare commented on homeassistant/config.py:
```diff
@@ -1023,12 +1035,36 @@ async def merge_packages_config(
+    Raises vol.Invalid if whole package config is invalid.
+    """
+
+    def _validate_package_definition(name: str, conf: Any) -> None:
```
This function can be moved out to the module level since it doesn't depend on any local context.

emontnemery replied:
Yes, but from a code readability perspective I think it's beneficial to keep it as an inner function / closure.

chammp replied:
See, I don't really care about one or the other; it's just unclear what to do now when reviewers don't agree on a topic 🤷 

I moved the function def up a couple of lines now; there _is_ precedence for this in the same file (several functions could have been defined as inner functions but weren't)

emontnemery reviewed (approved):
Thanks @chammp 👍 

This is a great first PR, we hope to see more coming 🚀

emontnemery committed:
Merge branch 'dev' into wrap_package_schema_validation_errors

emontnemery commented:
The test failures are unrelated, but we can let CI run once to get code coverage of the new code in codecov.

chammp committed:
Move inner function to module level

MartinHjelmare commented on homeassistant/helpers/check_config.py:
```diff
@@ -170,9 +170,16 @@ async def _get_integration(
+            hass, config, core_config.get(CONF_PACKAGES, {}), _pack_error
+        )
+    except vol.Invalid as err:
+        result.add_error(
```
Looks like this case is untested.

chammp replied:
Hardly without changing the flow of this method. The problem here is that `merge_packages_config` raises `Invalid` only if `PACKAGES_CONFIG_SCHEMA(packages)` inside that function fails. This will not happen in `async_check_ha_config_file` because a couple lines above, `core_config = CORE_CONFIG_SCHEMA(core_config)` already transitively performs the same checks that would lead to a failure during merge. (And knowing about the double checks is merely an implementation details that one cannot necessarily derive from the function signature and docs, so in any case I would leave the exception handler in place).


We could merge the two exception handlers into one, like so:

```python
    # Extract and validate core [homeassistant] config
    try:
        core_config = config.pop(CONF_CORE, {})
        core_config = CORE_CONFIG_SCHEMA(core_config)
        result[CONF_CORE] = core_config

        # Merge packages
        await merge_packages_config(
            hass, config, core_config.get(CONF_PACKAGES, {}), _pack_error
        )
    except vol.Invalid as err:
        result.add_error(
            format_schema_error(hass, err, CONF_CORE, core_config),
            CONF_CORE,
            core_config,
        )
        core_config = {}

    core_config.pop(CONF_PACKAGES, None)
```

I like the previous version slightly more as I find it easier to reason about. YMMV, of course. What's your opinion? Yes, no?

MartinHjelmare replied:
If we can't reach that code we could raise RuntimeError add a comment. @emontnemery may have an opinion.

chammp replied:
Nevermind, I merged both exception handlers now as described above.

chammp committed:
Merge exception handlers

Merge exception handlers for config schema validation and package merge
to avoid untested code branches

chammp committed:
Merge branch 'dev' into wrap_package_schema_validation_errors

chammp commented on homeassistant/helpers/check_config.py:
```diff
@@ -161,6 +161,12 @@ async def _get_integration(
+        await merge_packages_config(
+            hass, config, core_config.get(CONF_PACKAGES, {}), _pack_error
+        )
+        core_config.pop(CONF_PACKAGES, None)
```
I'm totally fine with separating both parts, raising an error and adding a comment; however I'm a little lost regarding that comment's actual wording 🤷. If you have an idea I'm game.

MartinHjelmare replied:
The top and bottom dict access can be moved out without further changes.

chammp replied:
Yes. That is for sure possible.

chammp committed:
Fix long lines and doc strings

chammp commented on homeassistant/config.py:
```diff
@@ -1023,12 +1041,31 @@ async def merge_packages_config(
     ] = _log_pkg_error,
 ) -> dict:
-    """Merge packages into the top-level configuration. Mutate config."""
+    """Merge packages into the top-level configuration. Ignores packages that cannot be setup. Mutates config.
```
I expected one of those linter bois to pick up issues like this... 😅

chammp commented on homeassistant/config.py:
```diff
@@ -1023,12 +1041,31 @@ async def merge_packages_config(
+                pack_name,
+                None,
+                config,
+                f"Invalid package definition '{pack_name}': {str(exc)}. Package will not be initialized",
```
Same as above - I expected a linter would pick up issues like this.

chammp committed:
More minor changes to exception handler

MartinHjelmare reviewed (approved):
Thanks!
